File: app/src/main/java/feelings/guide/answer/Answer.java
package feelings.guide.answer;

import androidx.annotation.NonNull;

import org.threeten.bp.LocalDateTime;

public class Answer {
    private long id;
    private final long questionId;
    private final LocalDateTime dateTime;
    private String answerText;

    public Answer(long questionId, LocalDateTime dateTime, String answerText) {
        this.questionId = questionId;
        this.dateTime = dateTime;
        this.answerText = answerText;
    }

    /**
     * Ctor for cases when ID is known beforehand. Ex: restoring deleted answer
     */
    Answer(long id, long questionId, LocalDateTime dateTime, String answerText) {
        this(questionId, dateTime, answerText);
        this.id = id;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getQuestionId() {
        return questionId;
    }

    public LocalDateTime getDateTime() {
        return dateTime;
    }

    public String getAnswerText() {
        return answerText;
    }

    public void setAnswerText(String answerText) {
        this.answerText = answerText;
    }

    // equals and hashCode for tests
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Answer answer = (Answer) o;

        if (questionId != answer.questionId) return false;
        if (!dateTime.equals(answer.dateTime)) return false;
        return answerText.equals(answer.answerText);
    }

    @Override
    public int hashCode() {
        int result = (int) (questionId ^ (questionId >>> 32));
        result = 31 * result + dateTime.hashCode();
        result = 31 * result + answerText.hashCode();
        return result;
    }

    @NonNull
    @Override
    public String toString() {
        return "Answer{" +
                "id=" + id +
                ", questionId=" + questionId +
                ", dateTime=" + dateTime +
                ", answerText='" + answerText + '\'' +
                '}';
    }
}

File: app/src/main/kotlin/feelings/guide/answer/Answer.kt
package feelings.guide.answer

import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

data class Answer(
        val questionId: Long,
        val dateTime: LocalDateTime,
        var answerText: String?
) {
    var id: Long = 0

    /**
     * Ctor for cases when ID is known beforehand. Ex: restoring deleted answer
     */
    internal constructor(id: Long, questionId: Long, dateTime: LocalDateTime, answerText: String)
            : this(questionId, dateTime, answerText) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        val o = other as Answer
        return questionId == o.questionId
                && dateTime.truncatedTo(ChronoUnit.MILLIS) == o.dateTime.truncatedTo(ChronoUnit.MILLIS)
                && answerText == o.answerText
    }

    override fun hashCode(): Int {
        var result = questionId.hashCode()
        result = 31 * result + dateTime.truncatedTo(ChronoUnit.MILLIS).hashCode()
        result = 31 * result + (answerText?.hashCode() ?: 0)
        return result
    }
}
